pytorch_lightning/trainer/trainer.py

```diff
@@ -943,7 +944,12 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
             self.call_hook("on_fit_end")
 
         # teardown
-        self._call_teardown_hook(model)
+        if self._distrib_type not in (
+            DistributedType.DDP_SPAWN,
+            DistributedType.TPU_SPAWN,
+            DistributedType.DDP_SHARDED_SPAWN,
+        ):
+            self._call_teardown_hook(model)
```

Comment:
same here: please add a comment for why spawn plugins are special-cased
@awaelchli @kaushikb11  - should we have a private helper function for determining if checking if the training type is spawned? would this be useful elsewhere?

Reply:
You can add a staticmethod like `interactive_compatible_types`

Reply:
can we do an instance check against the spawn plugin from which all these other plugins inherit?

Reply:
@awaelchli @kaushikb11 this could be a good follow up. the usage of these enums (device type, distributed type, amp, grad clip algorithm) will run into issues if/when users supply their own implementations. instance checks will enforce an inheritance structure for users to follow. i wonder if explicit properties per accelerator strategy plugin is another option

Reply:
Interesting.

For eg., a User implements their own `tpu_sharded_spawn` plugin. It would cause issues to the User, should we have functionality for Users to register or update attributes in Enums?

Yes, having explicit properties for accelerator strategy plugins is one option, but I like the idea of having it decoupled in Enums rather than having it with plugins.

Reply:
@ananthsub @kaushikb11 Those are good points. In this case, is it better to replace the enum check with `isinstance(self.training_type_plugin, ParallelPlugin)` as suggested by @awaelchli then?

Reply:
for now, i think @carmocca 's suggestion for `interactive_compatible_types` is good to proceed with this fix.
we can file a separate issue to audit the lightning codebase for how/where DistributedType is used